core/state-machine/src/backend.rs

```diff
@@ -239,6 +249,36 @@ impl<H: Hasher> Backend<H> for InMemory<H> where H::Out: HeapSizeOf {
 		self.inner.get(&Some(storage_key.to_vec())).map(|map| map.keys().for_each(|k| f(&k)));
 	}
 
+	fn full_storage_root<I>(&self, delta: I) -> (H::Out, Self::Transaction)
+	where
+		I: IntoIterator<Item=(Vec<u8>, Option<Vec<u8>>)>,
+		<H as Hasher>::Out: Ord,
+	{
+		let existing_pairs = self.inner.get(&None).into_iter().flat_map(|map| map.iter().map(|(k, v)| (k.clone(), Some(v.clone()))));
+		let transaction: Vec<_> = delta.into_iter().collect();
+		let mut map_input = existing_pairs.chain(transaction.iter().cloned())
+			.collect::<HashMap<_, _>>();
+
+		// first add child root to delta
+		for (storage_key, _existing_pairs) in self.inner.iter() {
+			if let Some(storage_key) = storage_key.as_ref() {
+				let child_root = self.child_storage_root(storage_key, ::std::iter::empty()).0;
+				map_input.insert(storage_key.clone(), Some(child_root));
+			}
+		}
+
+
+		let root = trie_root::<H, _, _, _>(map_input
+			.into_iter()
+			.filter_map(|(k, maybe_val)| maybe_val.map(|val| (k, val)))
+		);
+
+		let full_transaction = transaction.into_iter().map(|(k, v)| (None, k, v)).collect();
```

Comment:
Probably dumb question - I have forgot everything about children storage already :) Shouldn't `full_transaction` include transactions from child storage(s)? I mean there are multiple calls to `child_storage_root()` above && returned child-storage-transaction is just ignored (we only use root part of these calls).

Reply:
I did look, it is done in state_machine ext level through  `storage_root` -> `child_storage_root_transaction` -> `sync_child_storage_root`

Reply:
It would totally make sense indeed, I am changing it (before maybe removing the function) :)